### DISCUSSION TITLE
AMQP-1.0 scenario with 1 connection for both publisher and consumer

### DISCUSSION THREAD

**Discussion by annieblomgren** [rabbitmq/omq#31] [Ideas]
Hi,

It would be nice if omq supported having one shared connection for both publisher and consumer using AMQP-1.0
I have tried to implement it, but don't have enough go knowledge.

Was reading up on the native AMQP-1.0 features in [this PR](https://github.com/rabbitmq/rabbitmq-server/pull/9022), and notices that:

> There is no need for separate AMQP 1.0 connections for publishers and consumers as recommended in our AMQP 0.9.1 implementation.

**Comment by mkuratczyk** [comment]
Can you elaborate on this? Some questions that immediately jump to mind:
1. What is the use cases in real life for this? I'd assume that the app would consume from one queue, perform some operation and then publish to a different queue to trigger the next part of some larger workflow. Is that what you have in mind?
2. Any suggestions on what the CLI commands could looks like to use this feature if it existed?
3. Do you just want to reduce the number of connections or are there things you want to do with this publishing and consuming connection that you can't test right now?

**Reply by annieblomgren** [reply]
1. For an IoT scenario, where we want to have as many devices connected to a rabbitmq instance as possible, but we don't want to waste resources on connections. A device will need to send data to a server on the publishing side, as well as receive updates from the server on the consuming side. 

2. Hm not sure so just brainstorming here, if it's going to be AMQP-1.0 specific maybe something like --sessions? 
Otherwise --channels? 

```
./omq amqp --publishers 100 --consumers 100 --sessions 2 \
           --publish-to 'sensor-%d' --consume-from 'sensor-%d'
```
In this example it would be 100 connections and 200 sessions. It probably needs some assumptions tho, like number of publishers are equal to number of consumers. 

3. Yes, just reduce the number of connections.

**Reply by annieblomgren** [reply]
[Just noticed that the benchmark blog post on AMQP-1.0 on RabbitMQ 4.0 is using 2 sessions per connection](https://www.rabbitmq.com/blog/2024/08/21/amqp-benchmarks#many-connections), when testing many connections.

**Reply by mkuratczyk** [reply]
It is a perfectly valid way of using RabbitMQ over AMQP-1.0 and certainly beneficial in cases where the same app both publishes and consumes. I don't mind having such an option in `omq` but I have no plans to work on this soon (unless I suddenly need it to test a specific scenario, but I don't expect that).
